Add abort API in gloo ProcessGroup Send/Recv Work

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30020 [rpc] Add local shutdown to process group agent
* **#29928 Add abort API in gloo ProcessGroup Send/Recv Work**

Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

rohan-varma committed:
Add abort API in gloo ProcessGroup Send/Recv Work

- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.
ghstack-source-id: 94004848

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

xush6528 reviewed (commented):
Looks good to me. Need @pietern 's look too.

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -73,6 +73,8 @@ class ProcessGroup {
 
     // Waits until request completes. Blocking operation.
     // Throws if the work completed with an exception.
+    // Returns false if the work is aboorted.
```
nit: aborted.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -363,6 +364,12 @@ void ProcessGroupGloo::SendWork::wait() {
+}
+
+void ProcessGroupGloo::SendWork::abort() {
+  std::unique_lock<std::mutex> lock(mutex_);
```
If a thread is blocked on `buffer_->waitSend()` in `wait()`, there is no way to abort since you need to hold the lock here? I'm wondering if this resolves the problem we're seeing, since I think the abort() call would just block forever and can't really interrupt `waitSend`.

xush6528 replied:
we would like to remove this lock, if

- no other thread can assign to buffer_
- abortWaitSend is thread safe to buffer_

rohan-varma replied:
Synced with @xush6528  about this, we should be able to remove the locks here, `unbound_buffer` from gloo should be thread safe. We think the tests in this diff due to the way the threads were scheduled

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -387,6 +395,12 @@ void ProcessGroupGloo::RecvWork::wait() {
+}
+
+void ProcessGroupGloo::RecvWork::abort() {
+  std::unique_lock<std::mutex> lock(mutex_);
```
Same issue with lock as `waitSend`

pritamdamania87 reviewed (changes requested):
I might be missing something, but it seems like the way the locks are setup, the `abort` method can't interrupt `wait`?

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -387,6 +394,11 @@ void ProcessGroupGloo::RecvWork::wait() {
+}
+
+void ProcessGroupGloo::RecvWork::abort() {
+  buffer_->abortWaitRecv();
```
We should not have to acquire a lock here since this function does not do any writes to class variables, and `buffer_` in gloo should be thread safe.

pritamdamania87 replied:
Does this mean we can also avoid holding the lock in `RecvWork::wait` and `SendWork::wait` when we call `buffer_->waitRecv` and `buffer_->waitSend`? We could just acquire the lock when we set `completed_` and `exception_`?

cc @pietern

rohan-varma replied:
This makes sense to me unless I'm missing something.

pietern replied:
The semantics for the work is that there is only a single logical user/waiter. The lock is there only to synchronize a background thread that executes the work and marks it as completed, not multiple waiters. For the send/recv work, that background thread is the actual Gloo I/O thread that marks a send/recv as completed, so for this case we could get rid of the locks entirely.

rohan-varma replied:
Filed https://github.com/pytorch/pytorch/issues/30164 so we don't forget about this.

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma commented:
@pritamdamania87 @pietern Would be great if you all could take a look at this PR.

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -351,10 +351,11 @@ ProcessGroupGloo::SendWork::SendWork(
 
-void ProcessGroupGloo::SendWork::wait() {
+bool ProcessGroupGloo::SendWork::wait() {
+  bool sendCompleted;
```
Initialize to false. Otherwise, its uninitialized value may be returned if `waitSend` throws.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -375,10 +381,11 @@ int ProcessGroupGloo::RecvWork::sourceRank() const {
 
-void ProcessGroupGloo::RecvWork::wait() {
+bool ProcessGroupGloo::RecvWork::wait() {
+  bool recvCompleted;
```
Dito.

pietern reviewed (commented):
There is an uninitialized variable that could wreak havoc. Otherwise looks good!

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add abort API in gloo ProcessGroup Send/Recv Work"


Original author: @xush6528 
- Add abort to `c10d::ProcessGroup::Work`.
- Change the return type of `c10d::ProcessGroup::Work::wait()` to boolean to indicate if the work is aborted after waiting.
- Add unit test for the correctness of abort.

Differential Revision: [D5685727](https://our.internmc.facebook.com/intern/diff/D5685727/)

[ghstack-poisoned]
